pkg/binding/bindingpolicy-resolution.go

```diff
@@ -283,57 +212,50 @@ func destinationsMatch(resolvedDestinations sets.Set[string], bindingDestination
+	gvkGvrMapper util.GvkGvrMapper) bool {
 	// check lengths
-	clusterScopedObjectsCount := 0
-	for _, clusterScopeDownsyncObjects := range bindingSpecWorkload.ClusterScope {
-		clusterScopedObjectsCount += len(clusterScopeDownsyncObjects.ObjectNames)
-	}
-
-	namespacedObjectsCount := 0
-	for _, namespaceScopeDownsyncObjects := range bindingSpecWorkload.NamespaceScope {
-		for _, objectsByNamespace := range namespaceScopeDownsyncObjects.ObjectsByNamespace {
-			namespacedObjectsCount += len(objectsByNamespace.Names)
-		}
-	}
-
-	if len(objectIdentifierToKeyMap) != clusterScopedObjectsCount+namespacedObjectsCount {
+	if len(objectIdentifierToKeyMap) != len(bindingSpecWorkload.ClusterScope)+len(bindingSpecWorkload.NamespaceScope) {
 		return false
 	}
 
 	// again we can check match by making sure all objects are mapped, since entries are unique and the length is equal.
 	// check cluster-scoped all exist
 	if !bindingClusterScopeIsMapped(bindingSpecWorkload.ClusterScope, objectIdentifierToKeyMap,
-		gvkGvrMapper) {
+		objectIdentifierToVersionMap, gvkGvrMapper) {
 		return false
 	}
 
 	// check namespace-scoped all exist
-	return namespaceScopeMatchesBindingSpec(bindingSpecWorkload.NamespaceScope, objectIdentifierToKeyMap,
-		gvkGvrMapper)
+	return bindingNamespaceScopeIsMapped(bindingSpecWorkload.NamespaceScope, objectIdentifierToKeyMap,
+		objectIdentifierToVersionMap, gvkGvrMapper)
 }
 
 // bindingClusterScopeIsMapped returns true if the cluster-scope
 // section in the binding spec all exist in the resolution.
-func bindingClusterScopeIsMapped(bindingSpecClusterScope []v1alpha1.ClusterScopeDownsyncObjects,
-	objectIdentifierToKeyMap map[string]*util.Key, gvkGvrMapper util.GvkGvrMapper) bool {
-	for _, clusterScopeDownsyncObjects := range bindingSpecClusterScope {
-		gvr := schema.GroupVersionResource(clusterScopeDownsyncObjects.GroupVersionResource)
+func bindingClusterScopeIsMapped(bindingSpecClusterScope []v1alpha1.ClusterScopeDownsyncObject,
+	objectIdentifierToKeyMap map[string]*util.Key, objectIdentifierToVersionMap map[string]string,
+	gvkGvrMapper util.GvkGvrMapper) bool {
+	for _, clusterScopeDownsyncObject := range bindingSpecClusterScope {
+		gvr := schema.GroupVersionResource(clusterScopeDownsyncObject.GroupVersionResource)
 		gvk, found := gvkGvrMapper.GetGvk(gvr)
 
 		if !found {
 			return false // if not found then not mapped and not in resolution
 		}
 
-		for _, objName := range clusterScopeDownsyncObjects.ObjectNames {
-			formattedKey := util.KeyFromGVKandNamespacedName(gvk, types.NamespacedName{
-				Namespace: metav1.NamespaceNone,
-				Name:      objName,
-			})
+		formattedKey := util.KeyFromGVKandNamespacedName(gvk, types.NamespacedName{
+			Namespace: metav1.NamespaceNone,
+			Name:      clusterScopeDownsyncObject.Name,
+		})
 
-			if _, found := objectIdentifierToKeyMap[formattedKey]; !found {
-				return false
-			}
+		if _, found := objectIdentifierToKeyMap[formattedKey]; !found {
+			return false
+		}
```

Comment:
If `!found` then the ResourceVersion test below is guaranteed to fail, right? That makes this test unnecessary.